File: frontend/src/api/jobs.ts
import { api } from "./client";
import type { CompanySuggestion, Job, JobCreate, JobPostingSuggestion, JobUpdate } from "../types/job";

// Get all jobs
export async function listJobs(): Promise<Job[]> {
    const res = await api.get<Job[]>("/jobs/");
    return res.data;
}

// Create a new job
export async function createJob(payload: JobCreate): Promise<Job> {
    const res = await api.post<Job>("/jobs/", payload);
    return res.data;
}

// Update a job
export async function updateJob(id: number, payload: JobUpdate): Promise<Job> {
    const res = await api.put<Job>(`/jobs/${id}`, payload);
    return res.data;
}

// Delete a job
export async function deleteJob(id: number): Promise<void> {
    await api.delete(`/jobs/${id}`);
}

// Search specific job
export async function searchJobs(params:{
    company?: string;
    title?: string;
    location?: string;
    status?: string;
    skip?: number;
    limit?: number;
    sort_by?: string;
    sort_desc?: boolean;
}): Promise<Job[]> {
   const res = await api.get<Job[]>("/jobs/search",{params});
   return res.data;
}

// Add multiple jobs
export async function createJobsBatch(jobs: JobCreate[]) {
    const res = await api.post<Job[]>("/jobs/batch", jobs);
    return res.data;
}

export async function searchCompanies(query: string): Promise<CompanySuggestion[]> {
    const res = await api.get<CompanySuggestion[]>("/jobs/companies", {
        params: { query },
    });
    return res.data;
}

export async function searchJobPostings(company: string, query = ""): Promise<JobPostingSuggestion[]> {
    const res = await api.get<JobPostingSuggestion[]>("/jobs/postings", {
        params: { company, query },
    });
    return res.data;
}

File: backend/app/routers/jobs.py
from fastapi import APIRouter, HTTPException, Depends
from sqlalchemy.orm import Session
from typing import Optional
from ..db.database import get_db
from ..crud import crud
from ..schemas import schemas
from ..models import models

router = APIRouter(prefix="/jobs", tags=["Jobs"])

#Create a job
@router.post("/", response_model=schemas.Job)
def create_job_route(job: schemas.JobCreate, db: Session = Depends(get_db)):
    return crud.create_job(db=db, job=job)

#Return all jobs with pagination
@router.get("/", response_model=list[schemas.Job])
def read_jobs(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    jobs = crud.get_jobs(db, skip=skip, limit=limit)
    return jobs

#Search job by criteria
@router.get("/search", response_model=list[schemas.Job])
def search_jobs(
    company: Optional[str] = None,
    title: Optional[str] = None,
    location: Optional[str] = None,
    status: Optional[str] = None,
    skip: int = 0,
    limit: int = 100,
    sort_by: str = "applied_date",
    sort_desc: bool = True,
    db: Session = Depends(get_db)): 
    if not hasattr(models.Job, sort_by):
        raise HTTPException(status_code=400, detail=f"Invalid sort field: {sort_by}")
    return crud.get_jobs_by_filters(
        db=db, 
        company=company, 
        title=title, 
        location=location, 
        status=status,
        skip=skip,
        limit=limit,
        sort_by=sort_by,
        sort_desc=sort_desc
    )

#Get single job
@router.get("/{job_id}", response_model=schemas.Job)
def read_job(job_id: int, db: Session = Depends(get_db)):
    job = crud.get_job_by_id(db, job_id)
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")
    return job

#Delete a job
@router.delete("/{job_id}", response_model=schemas.Job)
def delete_job_route(job_id: int, db: Session = Depends(get_db)):
    db_job = crud.get_job_by_id(db, job_id)
    if not db_job:
        raise HTTPException(status_code=404, detail = "Job not found")
    return crud.delete_job(db=db, job_id=job_id)

#Update a job
@router.put("/{job_id}", response_model=schemas.Job)
def update_job(job_id: int, job_update: schemas.JobUpdate, db: Session = Depends(get_db)):
    job = crud.update_job(db=db, job_id=job_id,job_update=job_update)
    if not job:
        raise HTTPException(status_code=404, detail= "Job not found")
    return job

#Upload jobs in bulk
@router.post("/batch", response_model=list[schemas.Job])
def create_jobs_batch(jobs: list[schemas.JobCreate], db: Session = Depends(get_db)):
    created_jobs = []
    for job in jobs:
        created = crud.create_job(db=db, job=job)
        created_jobs.append(created)
    return created_jobs
